internal/provider/bms/bms_types.go

```diff
@@ -0,0 +1,139 @@
+package bms
+
+import (
+	"context"
+
+	timeoutsD "github.com/hashicorp/terraform-plugin-framework-timeouts/datasource/timeouts"
+
+	supertypes "github.com/FrangipaneTeam/terraform-plugin-framework-supertypes"
+
+	v1 "github.com/orange-cloudavenue/cloudavenue-sdk-go/v1"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/pkg/utils"
+)
+
+type (
+	bmsModelDatasource struct {
+		ID       supertypes.StringValue                                   `tfsdk:"id"`
+		Timeouts timeoutsD.Value                                          `tfsdk:"timeouts"`
+		Env      supertypes.SetNestedObjectValueOf[bmsModelDatasourceEnv] `tfsdk:"env"`
+	}
+
+	bmsModelDatasourceEnv struct {
+		Network supertypes.SetNestedObjectValueOf[bmsModelDatasourceNetwork] `tfsdk:"network"`
+		BMS     supertypes.SetNestedObjectValueOf[bmsModelDatasourceBMS]     `tfsdk:"bms"`
+	}
+
+	bmsModelDatasourceNetwork struct {
+		VLANID supertypes.StringValue `tfsdk:"vlan_id"`
+		Subnet supertypes.StringValue `tfsdk:"subnet"`
+		Prefix supertypes.StringValue `tfsdk:"prefix"`
+	}
+
+	bmsModelDatasourceBMS struct {
+		Hostname          supertypes.StringValue                                             `tfsdk:"hostname"`
+		BMSType           supertypes.StringValue                                             `tfsdk:"bms_type"`
+		OS                supertypes.StringValue                                             `tfsdk:"os"`
+		BiosConfiguration supertypes.StringValue                                             `tfsdk:"bios_configuration"`
+		Storage           supertypes.SingleNestedObjectValueOf[bmsModelDatasourceBMSStorage] `tfsdk:"storage"`
+	}
+
+	bmsModelDatasourceBMSStorage struct {
+		Local  supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"local"`
+		System supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"system"`
+		Data   supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"data"`
+		Shared supertypes.ListNestedObjectValueOf[bmsModelDatasourceBMSStorageGen] `tfsdk:"shared"`
+	}
+
+	bmsModelDatasourceBMSStorageGen struct {
+		Size         supertypes.StringValue `tfsdk:"size"`
+		StorageClass supertypes.StringValue `tfsdk:"storage_class"`
+	}
+)
+
+// NewbmsModelDatasource returns a new bmsModelDatasource.
+func NewBMSModelDatasource(ctx context.Context) *bmsModelDatasource {
+	return &bmsModelDatasource{
+		ID:       supertypes.NewStringNull(),
+		Timeouts: timeoutsD.Value{},
+		Env:      supertypes.NewSetNestedObjectValueOfNull[bmsModelDatasourceEnv](ctx),
+	}
+}
+
+// SetID sets the ID.
+func (m *bmsModelDatasource) SetID(id *string) {
+	m.ID.Set(*id)
+}
+
+// SetNetwork sets the Network of BMS listed.
+func (m *bmsModelDatasource) SetNetwork(ctx context.Context, bms *v1.BMS) (net []*bmsModelDatasourceNetwork, err error) {
```

Comment:
The function name doesn't seem appropriate to me. Indeed we do not "Set" values ​​in the struct `bmsModelDatasource` but we construct a terraform compatible object.  

I propose to rename it to `NetworkToTerraform` and this function is not method of `bmsModelDatasource`. 

The `ctx` is unused. 

```suggestion
func NetworkToTerraform(bms *v1.BMS) (net []*bmsModelDatasourceNetwork, err error) {
```


Suggest simplified func 
```go 
// NetworkToTerraform converts the network information to Terraform format.
func NetworkToTerraform(bms *v1.BMS) (net []*bmsModelDatasourceNetwork) {

	for _, network := range bms.GetNetworks() {
		var x bmsModelDatasourceNetwork

		x.VLANID.Set(network.VLANID)
		x.Subnet.Set(network.Subnet)
		x.Prefix.Set(network.Prefix)
		net = append(net, &x)
	}

	return
}
```

Reply:
Fix